std/typecons.d

```diff
@@ -2094,6 +2100,343 @@ enum isTuple(T) = __traits(compiles,
+    assert(x == 3);
+    // assert(marray[0].isInit); // Supposed to hold (see Limitations) ...
+    assert(marray[0] == 3); // ... instead of this.
+    // Warning: You should not rely on this behavior.
+
+    // Moving and Tuples.
+    auto mtuple = tuple(Move(2), Move(3));
+    tie[ x, $ ] = move(mtuple); // moved lvalue ...
+    assert(x == 2);
+    assert(mtuple[0].isInit); // ... from here
+    assert(mtuple[1].isInit); // destroyed
+    // Tuples do not support piecewise move extraction
+}
+
+/// `tie` supports non copyable and must-initialize types.
+pure nothrow @nogc @safe
+unittest
+{
+    import std.algorithm.mutation : move;
+
+    static struct NoCopy
+    {
+        int payload; // some payload
+        ~this() { } // destructor to observe moving
+        this(int value) { payload = value; } // necessary because must-initialize
+        bool opEquals(int value) { return payload == value; }
+        bool isInit() { return payload == 0; }
+        void opOpAssign(string op : "+")(NoCopy rhs) { payload += rhs.payload; }
+    @disable:
+        this(this); // make it non copyable
+        this(); // make it must-initialize
+    }
+
+    NoCopy nc = NoCopy(1), nc2 = NoCopy(3);
+
+    // This is copying assignment and therefore not allowed:
+    static assert(!__traits(compiles,
+    {
+        tie[ nc ] = tie[ nc ];
+    }));
+
+    // These are allowed moving assignments:
+    // right-hand side is rvalue
+    tie[ nc ] = tie[ NoCopy(2) ];
+    assert(nc == 2);
+    // right-hand side moved explicitly
+    tie[ nc ] = tie[ move(nc2) ];
+    assert(nc == 3);
+    // content of nc2 reset to init
+    assert(nc2.isInit);
+
+    nc2 = NoCopy(1);
+    // This is not allowed copying assignment ...
+    static assert(!__traits(compiles,
+    {
+        nc += nc2;
+    }));
+    // ... so it neither works in `tie`.
+    static assert(!__traits(compiles,
+    {
+        tie[ nc ] += tie[ nc2 ];
+    }));
+    // You need move-assignment.
+    tie[ nc ] += tie[ move(nc2) ];
+    assert(nc.payload == 4 && nc2.isInit);
+    nc2 = NoCopy(3);
+
+    // This is in no way real swapping:
+    static assert(!__traits(compiles,
+    {
+        tie[ nc, nc2 ] = tie[ nc2, nc ];
+    }));
+    // This roughly is, ...
+    tie[ nc, nc2 ] = tie[ move(nc2), move(nc) ];
+    assert(nc == 3 && nc2 == 4);
+    // ... but ...
+    import std.algorithm.mutation : swap;
+    swap(nc, nc2);
+    assert(nc == 4 && nc2 == 3);
+    // ... is preferred.
+
+    // Also works when NoCopy is inside a static array.
+    static NoCopy[2] f() { return [ NoCopy(1), NoCopy(2) ]; }
+    tie[ nc ] = cast(NoCopy[1]) [ move(nc2) ];
+    assert(nc == 3 && nc2.isInit);
+    tie[ nc, nc2 ] = f();
+    assert(nc == 1 && nc2 == 2);
+
+    NoCopy[2] ncarray = [ NoCopy(1), NoCopy(2) ];
+    static assert(!__traits(compiles, {
+        tie[ nc, $ ] = ncarray; // this copies!
+    }));
+    // These are supposed to work; see Limitations.
```

Comment:
Hmm, if we have nothing against them working, why unittest that they won't?

Reply:
As soon as the bug is fixed, they won't compile and the documentation can be adapted.
If that's not intended, I'll remove it.

Reply:
Good rationale.